### PR TITLE
Create VM Wizard - add ability to tab through edit fields in Storage and Networking steps 

### PR DESCRIPTION
**Pull request by bamsalem**
This PR added support in navigation between edit fields using keyboard:
1.Tab - move to the next field.
2. Shift+Tab - move to the previous field.

I used React refs to control the fields' focus.

tabThroughExample-2020-07-28_12 01 37

Fixes: #1248

### CODE REVIEW

**Comment by bamsalem** [comment]
Thanks Radek,
I checked your suggestion and I found that it not render the table and we can tab through the fields, and it's the purpose of this PR,
but it creates another issue, it don't render the table even after saving:
image
in the screenshot above you can see that it render the "Create Disk" button and we can see that it's enabled.
It's simpler solution, but it create another issue that we can't accept,
I tried to change your solution and solve the new issue but I didn't succeed to find a way to control when it will render and when not, so I'm not sure but I think that is not the solution.
screenshot from the dev tools:
image
WDYT?

**Comment by bamsalem** [comment]
@hstastna , I found why it happens:
https://github.com/oVirt/ovirt-web-ui/blob/ed40eb69a9fce4f1655d2db21e81af36eeaaf12d/src/components/CreateVmWizard/steps/Storage.js#L445-L462
In line 447 there is an edge case that zero is passed as a string so in the next render the 0 value stored as a string and that is what makes the warning,
It does not matter so much because I dealt with it in issue #1221 and it's not related to this issue, so I didn't include that changes here...

**Comment by hstastna** [comment]
Thank you very much, for the explanation, Ben! :)

**Comment by rszwajko** [comment]
> It's simpler solution, but it create another issue that we can't accept,

@bamsalem 
Save action should have an effect on the props used  to render the table. Probably sth similar to the way "Create disk" works.
IMHO lack of tabbing is a bug and should be fixed as such (no new framework for tabbing is required)

**Comment by bamsalem** [comment]
Hi Guys,
Thanks to @rszwajko  suggestion, I tried to handle this issue as a bug and not to create a new framework as it was handled in my prev fix.
So I found that the table re rendered because in every render() the <Table.PfProvider> component receive as a component prop a different reference and that what makes the re-rendering. So I find a way to control when the table will re-render and when it won't.
Now we can control the rendering of the table what enable us to tab through the fields.
After this PR will be approved we can make all the validations behave like in editing existing VM because we can use onChange events instead onBlur (I checked it in one field).
WDYT?
@sjd78  @sgratch @hstastna

**Comment by hstastna** [comment]
I think the problem with drop downs and changing selection was already mentioned in some other PR (or issue?) and as I remember it was not easy to fix it. The problem with Save button may be related. So I haven't mentioned those problems here. But yeah, we shouldn't forget about them :)

The problem with bootable flag, well... I haven't found this problem until Radek has mentioned it. I was not able to reproduce it. But now I found it's tricky. The error doesn't occur when pressing space key, but later - when clicking on Save button, and only when there's already some other (and not a template) bootable Disk. Also it does not matter if using keyboard or mouse. In addition, this problem is present also on master branch. I am pretty sure that this worked fine when merging bootable PR so some other very recent change had to cause it.

**Comment by sjd78** [comment]
> Problems I've encountered:
> 
> 1. dropdowns can be opened using keyboard (Enter key) but I was not able to change selection

See #585

**Comment by bamsalem** [comment]
>I think we can use shorter
this.components = this.components || components
instead of
this.components = this.components ? this.components : components

Otherwise LGTM :)


@hstastna done.

@sjd78 can we move the PR to QE?

**Review by hstastna — APPROVED** [review]

**Comment by bamsalem** [comment]
@hstastna @sgratch @sjd78 @rszwajko 
Hi guys,
I found  that this PR create another issue , it allows us using onChange event is better for validations (the validation is "on the go" and not on submit or onBlur), but it creates another issues like the issue I'll present here, important to note that the issue is in both of the steps: "Networking" & "Storage":
nicsResortingIssue-2020-08-25_09 35 27
like you can see the disks re-sorted during the editing and it could be annoying.
I'm not sure if it's related to this PR, or it should be dealt in another PR.
this PR's changes included in those PRs: #1219 #1230  #1221 
WDYT?

**Comment by hstastna** [comment]
@yfrimanm @matthewcarleton Is it a problem? https://github.com/oVirt/ovirt-web-ui/pull/1261#issuecomment-679844631

**Comment by yfrimanm** [comment]
Yes https://github.com/hstastna I totally agree it is a problem.

**Comment by hstastna** [comment]
IMHO, I would place the new item at the bottom of the table, as it was before, when creating it, and after the creation is done (after the item is saved) to sort all the items in the table normally.

**Comment by yfrimanm** [comment]
It does make sense to me as well https://github.com/hstastna

**Comment by sjd78** [comment]
@bamsalem, @hstastna - A simple way to handle things is to not update the fields that are used to sort until the save checkbox is checked.  So, keep state for the edit fields and then when the edit is done, run the doUpdate().

**Comment by bamsalem** [comment]
@hstastna @sjd78 @rszwajko @sgratch I pushed commit which resolves all sorting issues:
creating disks: always sorted to the bottom.
editing disks: sorted them according to the name before editing, and updateing them after saving and sorts them accordingly.
WDYT?

**Review by hstastna — APPROVED** [review]
LGTM.

File: src/components/CreateVmWizard/steps/Networking.js [hunk 1]
```diff
@@ -383,20 +384,26 @@ class Networking extends React.Component {
     const vnicList = createVNicProfileList(vnicProfiles, { dataCenterId, cluster })
     const enableCreate = vnicList.length > 0 && Object.keys(this.state.editing).length === 0
 
-    const nicList = nics
+    const nicList = [...nics]
```
Comment (sjd78) [inline src/components/CreateVmWizard/steps/Networking.js:387]:
```suggestion
    const nicList = nics.slice(0)
```

It is interesting to compare the various ways to create a shallow copy of an array.  Looks like slice(0) is the fastest across browsers.  I ran this in FF and Chrome:
  - https://measurethat.net/Benchmarks/Show/9356/0/array-slice-vs-slice0-vs-spread-operator

File: src/components/CreateVmWizard/steps/Storage.js [hunk 2]
```diff
@@ -558,9 +559,15 @@ class Storage extends React.Component {
     const filteredStorageDomainList = createStorageDomainList(storageDomains, dataCenterId)
     const enableCreate = storageDomainList.length > 0 && !this.isEditingMode()
 
-    const diskList = disks
+    const diskList = [...disks]
```
Comment (sjd78) [inline src/components/CreateVmWizard/steps/Storage.js:562]:
```suggestion
    const diskList = disks.slice(0)
```

File: src/components/CreateVmWizard/steps/Networking.js (hunk 1 above)
Comment (sjd78) [reply src/components/CreateVmWizard/steps/Networking.js:387]: On further thought, the spread operator is ok in this scenario.  @rszwajko is correct, the spread does look better than `.slice(0)`.

Change is not needed.

File: src/components/CreateVmWizard/steps/Storage.js (hunk 2 above)
Comment (sjd78) [reply src/components/CreateVmWizard/steps/Storage.js:562]: Sticking with the spread is ok.

**Comment by sjd78** [comment]
Note: This PR should merge before #1219, #1221, and maybe #1233 since the onChange handlers now work instead of relying on onBlur.

File: src/components/CreateVmWizard/steps/Networking.js (hunk 1 above)
Comment (hstastna) [reply src/components/CreateVmWizard/steps/Networking.js:387]: Now it feels like looking better is more than being faster. Could you provide some more explanation why we prefer the spread operator?

Comment (rszwajko) [reply src/components/CreateVmWizard/steps/Networking.js:387]: > being faster

.... in given test scenario on given browser version. Performance tweaks are tricky ;/

>  we prefer

well... I prefer:)

> some more explanation 

IMHO this syntax is very concise(short) and better shows the intention of the the programmer(copy of the whole array). 
As for more objective reasons - first `slice` can be easily switched by mistake with `splice`. 
Second `slice` main purpose is to copy portion of an array - making a copy of the whole array is supported but may be confusing.

Comment (hstastna) [reply src/components/CreateVmWizard/steps/Networking.js:387]: Okay, Thanks very much for the explanation, Radek! :)

**Comment by pnovotny** [comment]
LGTM

Tab and Shift+Tab through the fields for NICs and disks works properly. 

Side note: having #585 working would make this feature more awesome. Currently, only values in text inputs and checkboxes can be set with keyboard.

**Pull request merged**